File: Challenge-002-Single-linked-list/python/main.py
#!python3
# Template for Python script
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SingleLinkedList:
    def __init__(self):
        self.head = None

    def InsertAtBeginning(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    def InsertAtEnd(self, data):
        node = Node(data)
        if not self.head:
            self.head = node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = node

    def DeleteNodeAtIndex (self,index):
        temp = self.head
        counter = 2
        if self.head:
            if index==0 and index==1:
                temp = temp.next
                self.head = temp
            else:
                while temp.next.next and counter<index:
                    temp = temp.next
                    counter+=1
                aux = temp.next
                temp.next = aux.next

    def Traverse(self):
        current = self.head
        result = []
        while current:
            result.append(current.data)
            current = current.next
        print(result)

if __name__ == "__main__":
    ListA = SingleLinkedList()
    ListA.InsertAtBeginning(3)
    ListA.InsertAtEnd(2)
    ListA.InsertAtEnd(4)
    ListA.InsertAtBeginning(5)
    ListA.DeleteNodeAtIndex(3)
    ListA.Traverse()

File: Challenge-002-Single-linked-list/rust/src/main.rs
#[derive(Debug)]
struct Node {
    data: i32,
    next: Option<Box<Node>>,
}

struct SingleLinkedList {
    head: Option<Box<Node>>,
}

impl SingleLinkedList {
    fn new() -> Self {
        SingleLinkedList { head: None }
    }

    fn insert_at_beginning(&mut self, data: i32) {
        let new_node = Box::new(Node {
            data,
            next: self.head.take(),
        });
        self.head = Some(new_node);
    }

    fn insert_at_end(&mut self, data: i32) {
        let mut new_node = Box::new(Node { data, next: None });
        match self.head.as_mut() {
            Some(mut current) => {
                while let Some(ref mut next) = current.next {
                    current = next;
                }
                current.next = Some(new_node);
            }
            None => {
                self.head = Some(new_node);
            }
        }
    }

    fn delete_node_at_index (&mut self, index: usize){
        if index == 0 || index==1 {
            if let Some(node) = self.head.take() {
                self.head = node.next;
            }
            return;
        }
        let mut current = self.head.as_mut();
        let mut i = 0;
        while let Some(node) = current {
            if i + 1 == index {
                if let Some(next_node) = node.next.take() {
                    node.next = next_node.next;
                }
                return;
            }
            current = node.next.as_mut();
            i += 1;
        }
    }

    fn traverse(&self) {
        let mut current = self.head.as_ref();
        let mut result = Vec::new();
        while let Some(node) = current {
            result.push(node.data);
            current = node.next.as_ref();
        }
        println!("{:?}", result);
    }
}

fn main() {
    let mut list = SingleLinkedList::new();
    list.insert_at_beginning(3);
    list.insert_at_end(2);
    list.insert_at_end(4);
    list.insert_at_beginning(5);
    list.delete_node_at_index(3);
    list.traverse(); // Output: [5, 3, 2, 4]
}